sdk/core/core/src/az_json_pointer.c

```diff
@@ -1,28 +1,38 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
-#include "az_span_reader_private.h"
+#include <az_contract_internal.h>
 #include <az_json.h>
 
 #include <_az_cfg.h>
 
 AZ_NODISCARD static az_result az_span_reader_read_json_pointer_char(
-    az_span_reader * self,
+    az_span * self,
     uint32_t * out) {
   AZ_CONTRACT_ARG_NOT_NULL(self);
+  int32_t reader_current_length = az_span_length(*self);
 
-  az_result_byte const result = az_span_reader_current(self);
+  // check for EOF
+  if (reader_current_length == 0) {
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  uint8_t const result = az_span_ptr(*self)[0];
   switch (result) {
-    case AZ_ERROR_EOF: {
-      return AZ_ERROR_ITEM_NOT_FOUND;
-    }
     case '/': {
       return AZ_ERROR_JSON_POINTER_TOKEN_END;
     }
     case '~': {
-      az_span_reader_next(self);
-      az_result_byte const e = az_span_reader_current(self);
-      az_span_reader_next(self);
+      // move reader to next position
+      AZ_RETURN_IF_FAILED(az_span_slice(*self, 1, -1, self));
+      // check for EOF
+      if (az_span_length(*self) == 0) {
+        return AZ_ERROR_EOF;
+      }
+      // get char
+      uint8_t const e = self->_internal.ptr[0];
```

Comment:
call az_span_ptr; do not access internal fields directly. Please scan all the code for this. Only the span code should access span's internal fields directly.

As a general rule, only the code that implements a type should access the type's internal fields directly, other code must not. Maybe you and/or Anton can scan the code for _internal and check that all uses are appropriate. If we can't replace the internal field access with a method, then we need to add a method - not keep the internal field access.